src/messaging/ExchangeContext.cpp

```diff
@@ -326,7 +326,13 @@ ExchangeContext::~ExchangeContext()
     VerifyOrDie(!IsAckPending());
 
     if (ReleaseSessionOnDestruction() && mSession)
-        mSession->AsSecureSession()->MarkForRemoval();
+    {
+        // Move the session out of the older for the removal, so it doesn't try
+        // to notify us in our destructor.
+        const auto & session = mSession.Get();
```

Comment:
```suggestion
        const Optional<SessionHandle> session = mSession.Get();
```

Get returns a temporary variable, I'm not confident that it is lives long enough. We should use a value to hold it.